R/read-gs.R

```diff
@@ -18,14 +19,225 @@
+#' @importFrom dplyr filter select slice rename_with rename bind_cols mutate_at
+#' @importFrom purrr discard
+#' @importFrom lubridate make_difftime
+#' 
+read_canvas_grades <- function(grades){
+  # Convert Canvas grade dataframe into standard format
+  #
+  # Takes in a dataframe uploaded from Canvas. 
+  # Returns a dataframe in our standardized format (ie like Gradescope).
+  #
+  #
+  
+  # first determine which columns are assignments
+  
+  # canvas categories will not be kept
+  
+  assignments <- stringr::str_extract(names(grades),
+                                      ".+\\s*\\(\\d+\\)") |>
+                purrr::discard(is.na)
+  
+  # coerce into numerical and convert excused assignments into set value
+  grades <- grades |>
+    dplyr::filter(!is.na(Student)) |>
+    dplyr::mutate_at(assignments, function(x) {
+        dplyr::if_else(x == "EX", NA, x) |>
+        as.numeric()
+      })
+  
+ 
+  # now reconfigure the points possible info
+  max_points <- dplyr::filter(grades, Student %in% c( "    Points Possible", "Points Possible")) |>
+    dplyr::select(all_of(assignments)) |>
+    dplyr::slice(rep.int(1, nrow(grades))) |>
+    dplyr::rename_with(
+        function(x){
+          paste0(x, " - Max Points")
+        }
+    )
+  
+  
+  # remove unneeded cols, add max_point column, and remove rows that are not students
+  grades <- grades |>
+    dplyr::select(c(all_of(assignments), "ID", "Student", 
+                    "SIS User ID", "Section")) |>
+    dplyr::bind_cols(max_points) |>
+    dplyr::filter(!(Student %in% c("    Points Possible", "Student, Test", "Points Possible", NA)))
+  
+  # add columns for submission time and lateness
+  
+  sub_cols <- paste0(assignments, " - Submission Time")
+  
+  late_cols <- paste0(assignments, " - Lateness (H:M:S)")
+  
+  grades[sub_cols] <- NA
+  
+  grades[late_cols] <- lubridate::make_difftime(NA)
+  
+  # extract student name data
+  
+  grades$`First Name` <- (stringr::str_match(grades$Student, 
+                                             ".+,\\s(.+)"))[, 2]
+  
+  grades$`Last Name` <- (stringr::str_match(grades$Student, 
+                                            "(.+),.+"))[, 2]
+  
+  # change names to match Gradescope convention
+  grades <- grades |>
+    dplyr::rename(SID = `SIS User ID`,
+                  Sections = Section) |>
+    # reorder columns to match Gradescope
+    dplyr::select(c("First Name", "Last Name", "SID", "Sections",
+                    as.vector(
+                            mapply(c,
+                                   assignments, 
+                                    paste0(assignments, " - Max Points"), 
+                                    sub_cols,
+                                    late_cols)
+                              )
+                    )
+                  ) |>
+    # remove "UID:" moniker and make SID numeric
+    dplyr::mutate_at("SID", function(x){
+      stringr::str_replace(x, "^UID:", "") |> 
```

Comment:
I think we should talk through how to work with UIDs because we don't want this to create issues when we send grades out... if we remove the "UID", we don't know that these aren't the Berkeley SIDs